examples/src/java/com/twitter/heron/examples/streamlet/ImpressionsAndClicksTopology.java

```diff
@@ -0,0 +1,197 @@
+    public String toString() {
+      return String.format("(adId; %s, impressionId: %s)",
+          adId,
+          impressionId
+      );
+    }
+  }
+
+  /**
+   * A POJO for incoming ad clicks (generated every 50 milliseconds).
+   */
+  private static class AdClick implements Serializable {
+    private static final long serialVersionUID = 7202766159176178988L;
+    private String adId;
+    private String userId;
+    private String clickId;
+
+    AdClick() {
+      this.adId = StreamletUtils.randomFromList(ADS);
+      this.userId = StreamletUtils.randomFromList(USERS);
+      this.clickId = UUID.randomUUID().toString();
+      LOG.info(String.format("Emitting click: %s", this));
+    }
+
+    String getAdId() {
+      return adId;
+    }
+
+    String getUserId() {
+      return userId;
+    }
+
+    @Override
+    public String toString() {
+      return String.format("(adId; %s, clickId: %s)",
+          adId,
+          clickId
+      );
+    }
+  }
+
+  /**
+   * All Heron topologies require a main function that defines the topology's behavior
+   * at runtime
+   */
+  public static void main(String[] args) throws Exception {
+    Builder processingGraphBuilder = Builder.createBuilder();
+
+    // A KVStreamlet is produced. Each element is a KeyValue object where the key
+    // is the impression ID and the user ID is the value.
+    Streamlet<AdImpression> impressions = processingGraphBuilder
+        .newSource(AdImpression::new);
+
+    // A KVStreamlet is produced. Each element is a KeyValue object where the key
+    // is the ad ID and the user ID is the value.
+    Streamlet<AdClick> clicks = processingGraphBuilder
+        .newSource(AdClick::new);
+
+    /**
+     * Here, the impressions KVStreamlet is joined to the clicks KVStreamlet.
+     */
+    impressions
+        // The join function here essentially provides the reduce function with a streamlet
+        // of KeyValue objects where the userId matches across an impression and a click
+        // (meaning that the user has clicked on the ad).
+        .join(
+            // The other streamlet that's being joined to
+            clicks,
+            // Key extractor for the impressions streamlet
+            impression -> impression.getUserId(),
+            // Key extractor for the clicks streamlet
+            click -> click.getUserId(),
+            // Window configuration for the join operation
+            WindowConfig.TumblingCountWindow(25),
+            // Join type (inner join means that all elements from both streams will be included)
+            JoinType.INNER,
+            // For each element resulting from the join operation, a value of 1 will be provided
+            // if the user IDs match between the elements (or a value of 0 if they don't).
+            (user1, user2) -> (user1.getAdId().equals(user2.getAdId())) ? 1 : 0
+        )
+        // The reduce function counts the number of ad clicks per user.
+        .reduceByKeyAndWindow(
```

Comment:
this reduce will not needed once you change the key in the join

Reply:
Okay, yeah, I see the issue now. Will change up the processing logic accordingly.